### ISSUE TITLE
Lint against redundant field names in struct literals

### ISSUE THREAD

**Issue by KodrAus** [rust-lang-nursery/rust-clippy#2244]
Hi there! I think it would be nice to lint (warn by default?) when a struct field is assigned to a variable with the same name in a structure literal, but the name is still fully written out:

```rust
let id = 1;

let data = Data {
    id: id,
//  ^~~~~~ could be simplified to just `id`
};
```

**Comment by TomasKralCZ** [comment]
I would like to try to implement this.
I'm new to clippy and this would be my first issue. As far as I understand, this would involve dealing with `syntax::ast::FieldPat`, right ?

**Comment by oli-obk** [comment]
> Okay, I somehow figured it out. I have one question: should we lint in this case

I'm not sure I understand. Do you have an example code?

**Comment by TomasKralCZ** [comment]
> I'm not sure I understand. Do you have an example code?

Sorry, I (wrongly) assmued, that
`let fizz = Buzz { bar: foo::bar };`

could be written as.
`let fizz = Buzz { foo::bar }`

Anyway, would you please take a look at my code ? [gist](https://gist.github.com/TomasKralCZ/7acd7456df9f8720ebdcc007023a1303)
I'm sure I reinvented the wheel at least 5 times in there, and I don't want to do that.

I also wanted to ask, if there is documentation for `utils`, because that would be pretty helpful.

**Comment by oli-obk** [comment]
@TomasKralCZ lgtm modulo some nits.

You can always just open a PR, even an unfinished one. That's easier to comment on ;)

> I also wanted to ask, if there is documentation for utils, because that would be pretty helpful.

uuuh. xD not really The best we have is an old version on docs.rs: https://docs.rs/clippy_lints/0.0.156/clippy_lints/utils/index.html

You can always do `cargo doc` and generate them fresh

**Comment by TomasKralCZ** [comment]
I ran the tests and noticed, that it's already implemented in `no_effect`.
` error: statement with no effect
   --> $DIR/no_effect.rs:85:5
    |
 85 |     FooString { s: s };
    |     ^^^^^^^^^^^^^^^^^^^
 
`.

**Comment by vityafx** [comment]
Could anyone point me please to the rustc rfc which proposed such changes in the compiler? I'd argue it is useful...

**Comment by zackmdavis** [comment]
@vityafx [RFC 1682](https://github.com/rust-lang/rfcs/blob/24683b3fb/text/1682-field-init-shorthand.md)

**Comment by chrisduerr** [comment]
This has been implemented in #2451, the issue can probably be closed. Even though there are still some false positives.

**Comment by MightyPork** [comment]
I believe this lint is a bit misguided, or at least there are downsides to this.
When you use the IDEA Rust plugin and refactor>rename the variable after applying this fix, you get a compile error.

```rust
struct Foo {
  bar: u32
}

let bar = 15;

let a = Foo {
  bar: bar
};
```
1.
![screenshot_20190212_144313](https://user-images.githubusercontent.com/2041118/52639484-ae13fa00-2ed4-11e9-8ce3-88f0243601ae.png)
2.
![screenshot_20190212_144321](https://user-images.githubusercontent.com/2041118/52639485-aeac9080-2ed4-11e9-9600-7c80f51dc67f.png)
3.
![screenshot_20190212_144347](https://user-images.githubusercontent.com/2041118/52639486-aeac9080-2ed4-11e9-9667-e4cde3ee7bc7.png)

**Comment by oli-obk** [comment]
To me that seems more like a bug in the refactor+rename tool than in the clippy lint.
